drv/lpc55-sprot-server/src/main.rs

```diff
@@ -223,117 +234,181 @@ impl Io {
         }
     }
 
+    // Wait for CSn to actually be de-asserted.
+    //
+    // This checks the actual gpio pin, not the SPI block saturating bit which
+    // may be set from a prior deassertion.
+    //
+    // If CSn is still asserted at this point, then we are desynchronized and so
+    // we return an error.
+    //
+    // See https://github.com/oxidecomputer/hubris/issues/1507 for why we do
+    // this.
+    fn wait_for_csn_deasserted(&mut self) -> Result<(), DesynchronizedError> {
+        let mut result = Ok(());
+        while self.gpio.read_val(CHIP_SELECT) != Value::One {
+            ringbuf_entry!(Trace::Desynchronized);
+            if result.is_ok() {
+                self.stats.desynchronized =
+                    self.stats.desynchronized.wrapping_add(1);
+            }
+            result = Err(DesynchronizedError);
+        }
+        result
+    }
+
+    // Wait for a request from the SP
     pub fn wait_for_request(
         &mut self,
         rx_buf: &mut [u8],
     ) -> Result<usize, IoError> {
-        self.wait_for_csn_asserted(false);
+        self.wait_for_csn_asserted();
 
-        // Go into a tight loop receiving as many bytes as we can until we see
-        // CSn de-asserted.
         let mut bytes_received = 0;
         let mut rx = rx_buf.iter_mut();
-        while !self.spi.ssd() {
+
+        // Check for the SOT bit on first fifo read to see if we are
+        // synchronized.
+        let mut first_read = true;
+
+        let mut read_fifo = || {
             while self.spi.has_entry() {
                 bytes_received += 2;
-                let read = self.spi.read_u16();
+                let (read, sot) = self.spi.read_u16_with_sot();
+                if first_read {
+                    first_read = false;
+                    if !sot {
+                        self.stats.desynchronized =
+                            self.stats.desynchronized.wrapping_add(1);
```

Comment:
Maybe `saturating_add` instead?

Reply:
I made this match all the other counters. I specifically did that because it seems more useful to know if a counter is moving than if it saturated at one point in time. Although, in either case we shouldn't really come close to hitting this.

Reply:
Okay, I don't feel strongly about this!